src/Dialect/ONNX/Transforms/QlinearConvPass.cpp

```diff
@@ -0,0 +1,327 @@
+static bool extractScalarFloatFromConst(mlir::Value v, float &out) {
+  auto def = v.getDefiningOp<ONNXConstantOp>();
+  if (!def)
+    return false;
+
+  mlir::Attribute raw;
+  if (def.getValue().has_value())
+    raw = *def.getValue();
+  else
+    raw = def.getValueAttr();
+
+  if (auto elts = raw.dyn_cast<mlir::ElementsAttr>()) {
+    for (auto apf : elts.getValues<llvm::APFloat>()) {
+      out = apf.convertToFloat();
+      return true;
+    }
+    return false;
+  }
+
+  return false;
+}
+
+static mlir::DenseElementsAttr extractDenseFloatFromConst(
+    ONNXConstantOp constBiasOp, mlir::Value biasFloatValue) {
+  mlir::DenseElementsAttr denseBiasF;
+
+  if (!constBiasOp.getValue().has_value())
+    return denseBiasF; // empty, caller must check
+
+  mlir::Attribute opt = *constBiasOp.getValue();
+
+  if (auto elts = opt.dyn_cast<mlir::ElementsAttr>()) {
+    // Collect floats
+    llvm::SmallVector<float, 32> floats;
+    floats.reserve(elts.getNumElements());
+    for (auto v : elts.getValues<llvm::APFloat>())
+      floats.push_back(v.convertToFloat());
+
+    // Ensure correct tensor type shape
+    auto desiredTy =
+        biasFloatValue.getType().dyn_cast<mlir::RankedTensorType>();
+    if (!desiredTy)
+      return mlir::DenseElementsAttr(); // invalid → return empty
+
+    // Build DenseElementsAttr
+    denseBiasF =
+        mlir::DenseElementsAttr::get(desiredTy, llvm::ArrayRef<float>(floats));
+  }
+
+  return denseBiasF;
+}
+
+static llvm::SmallVector<mlir::Attribute, 32> createBiasI32Attrs(
+    mlir::DenseElementsAttr denseBiasF, double xScaleS, double wScaleS,
+    mlir::PatternRewriter &rewriter) {
+  llvm::SmallVector<mlir::Attribute, 32> biasI32Attrs;
+
+  double denom = static_cast<double>(xScaleS) * static_cast<double>(wScaleS);
+  if (denom == 0.0)
+    return biasI32Attrs; // empty, caller should check
+
+  biasI32Attrs.reserve(denseBiasF.getNumElements());
+  for (auto apf : denseBiasF.getValues<llvm::APFloat>()) {
+    double f = apf.convertToFloat();
+    double qd = std::nearbyint(f / denom);
+    int64_t qi = static_cast<int64_t>(qd);
+
+    // clamp to int32 range
+    if (qi > std::numeric_limits<int32_t>::max())
+      qi = std::numeric_limits<int32_t>::max();
+    if (qi < std::numeric_limits<int32_t>::min())
+      qi = std::numeric_limits<int32_t>::min();
+
+    biasI32Attrs.push_back(
+        rewriter.getI32IntegerAttr(static_cast<int32_t>(qi)));
+  }
+
+  return biasI32Attrs;
+}
+
+/// Pattern that matches Conv fed by DequantizeLinear(s) and consumed by
+/// QuantizeLinear.
+struct ConvToQLinearConvPattern : public OpRewritePattern<ONNXConvOp> {
+  using OpRewritePattern<ONNXConvOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      ONNXConvOp convOp, PatternRewriter &rewriter) const override {
+    Location loc = convOp.getLoc();
+
+    // Conv input X must come from DequantizeLinear ----
+    auto dqInputOp =
+        dyn_cast_or_null<ONNXDequantizeLinearOp>(convOp.getX().getDefiningOp());
```

Comment:
Nit, no change needed:
You can write this more compact as:
```suggestion
        convOp.getX().getDefiningOp<ONNXDequantizeLinearOp>();
```